### PR TITLE
[WIP] Some optimizations in dl4j

### PR DESCRIPTION
**Pull request by alexanderst**
## What changes were proposed in this pull request?

Optimization for model listener.

## How was this patch tested?

Unit tests.

### CODE REVIEW

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java [hunk 1]
```diff
@@ -37,7 +37,48 @@
...
+        }
+
+        public void reallocateGuard() {
+            if (position >= bufferLength * bucketNumber) {
+
+                int fullLength = bufferLength * bucketNumber;
+                int[] newIndexes = new int[fullLength * 2];
+                double[] newScores = new double[fullLength * 2];
+
+                if (position > Integer.MAX_VALUE) {
```
Comment (raver119) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java]: How exactly this condition is going to be met I wonder? Integer variable holding result bigger than Integer.MAX_VALUE

**Review by raver119 — DISMISSED** [review]
Fix that position bug, add javadoc, and should be good.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java (hunk 1 above)
Comment (alexanderst) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java]: Replaced with ==.

**Review by raver119 — CHANGES_REQUESTED** [review]
LGTM, but some kind of testing would be nice

**Review by raver119 — APPROVED** [review]

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java [hunk 2]
```diff
@@ -37,7 +37,74 @@
...
+            scores.add(new double[BUCKET_LENGTH]);
+        }
+
+        public List<long[]> getIndexes() {
+            return indexes;
+        }
+
+        public List<double[]> getScores() {
+            return scores;
+        }
```
Comment (AlexDBlack) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java:64]: For better usability, let's provide a method like indexes() and scores() that returns long[] and double[] with only the real data, without the extra preallocated bit at the end.
With the current methods, in almost all cases in practice, users will get a list of length 1, with an array of unknown "real" length - i.e., long[]/double[] with a bunch of 0s at the end.

On the off chance that users have more than Integer.MAX_VALUE total entries (should almost never happen in practice) we can just say "use getIndexes() and getPosition()"

**Review by AlexDBlack — CHANGES_REQUESTED** [review]
Overall good, just one minor UX improvement required :+1:

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java (hunk 2 above)
Comment (alexanderst) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/listeners/CollectScoresIterationListener.java:64]: Methods and test added.

**Review by AlexDBlack — APPROVED** [review]

**Pull request merged**